### PR TITLE
[#2161] One-Stop Config File for Code Portfolio

### PR DESCRIPTION
**Pull request by asdfghjkxd**
Part of #2161 
Part of #2170 

### Proposed commit message

```
Currently, `report-config.json` is underutilised (it is only used for
defining report titles). However, considering the code portfolio
feature, we need to implement a config file that allows users to
quickly set up and deploy it to their favourite platform to host
websites.

Let's move to repurpose `report-config.json` into a YAML file that
users can use it to set up configs to build their code portfolio more
easily.
```

### Other information

The code written for this feature is stored on the main repository to allow all team members to work on this concurrently. 

Additionally, the following PRs/Tasks need to be completed before we can merge this PR:

- [ ] Migrate CI/CD from Java 8 to Java 11 #2183

### CODE REVIEW

File: docs/ug/report-config.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+repoUrl: github.com/user/repo
+reportTitle: RepoSense Report
+authorDisplayName: Sample Author
+authorGithubId: Sample Author Github ID
+branches:
+  - name: Branch 1
+    blurb: Blurb 1
```
Comment (damithc) [inline docs/ug/report-config.yaml:7]: Blurb should allow a few paragraph of markdown text. So, yml might not be the right format for this config file.

File: docs/ug/report-config.yaml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+repoUrl: github.com/user/repo
+reportTitle: RepoSense Report
+authorDisplayName: Sample Author
```
Comment (damithc) [inline docs/ug/report-config.yaml:3]: Not needed

File: docs/ug/report-config.yaml [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+authorGithubId: Sample Author Github ID
+branches:
+  - name: Branch 1
+    blurb: Blurb 1
+  - name: Branch 2
+    blurb: Blurb 2
+startDate: 2020-01-01
+endDate: 9999-12-31
```
Comment (damithc) [inline docs/ug/report-config.yaml:11]: See https://github.com/reposense/RepoSense/issues/2161#issuecomment-2022776952 for an example that seems intuitive from the user POV

**Review by damithc — COMMENTED** [review]
Added some comments about the file format.

File: docs/ug/report-config.yaml (hunk 1 above)
Comment (asdfghjkxd) [reply docs/ug/report-config.yaml:7]: @damithc There are ways to incorporate multi-line strings into YAML files [here](https://yaml-multiline.info/), we could perhaps include this into the docs to let users know how they can tailor the blurbs to suit their use cases (multi-paragraph vs long single paragraph text)

File: docs/ug/report-config.yaml (hunk 2 above)
Comment (asdfghjkxd) [reply docs/ug/report-config.yaml:3]: @damithc Sure thing, will remove these from the YAML file and other associated classes in a future commit.

File: docs/ug/report-config.yaml (hunk 3 above)
Comment (asdfghjkxd) [reply docs/ug/report-config.yaml:11]: @damithc Just wanted to clarify, do you mean that we should include the configs within the markdown files itself, and have the backend parse the markdown files and then generate the reports, or that the overall order/style/keys in the YAML file should match what was used in the example provided?

Comment (damithc) [reply docs/ug/report-config.yaml:11]: @asdfghjkxd I wasn't specifically using markdown either. It is just a custom format that makes sense from the user's point of view. I'm fine if something close can be achieved by YAML or some other commonly-used format, provided it makes sense from the user (i.e., the user shouldn't have to do a lot of extra work just to conform to the file format)

**Comment by damithc** [comment]
@asdfghjkxd Let's settle the file format first. It's best to design the user view of the feature before we start implementing. Giving a full-fledged example config file is a good way to see the user POV of this feature.
For example, you can start with a YAML example that is equivalent of https://github.com/reposense/RepoSense/issues/2161#issuecomment-2022776952 and we can see if that is friendly enough for the user, and flexible enough for enhancements that we might add in future.
We can discuss at #2161

File: docs/ug/report-config.yaml (hunk 3 above)
Comment (asdfghjkxd) [reply docs/ug/report-config.yaml:11]: @damithc I see, sure thing, we will try to design a user-friendly and convenient config file format and update this PR and parent issue with any new findings/improvements!

**Comment by asdfghjkxd** [comment]
@damithc Thank you Prof for the feedback and guidance regarding this issue! We will look into improving the current config file format and update this PR and #2161 with any new findings or points of discussion!

**Comment by asdfghjkxd** [comment]
Do note that this PR should not be merged until we have successfully migrated our CI/CD from Java 8 to Java 11. This PR requires #2183.

**Comment by asdfghjkxd** [comment]
This PR has been split into 2 separate PRs (#2191 and #2192), and will hence be closed.

**Pull request closed without merging**